WSTEAMA-1558 - Script switch refactor

amoore108 opened the pull request:
Resolves JIRA https://jira.dev.bbc.co.uk/browse/WSTEAMA-1558

Overall changes
======
- Refactors the Script switch button to remove `scriptSwitchId` logic, as well as logic tied to `react-router-dom`
- This also allows the script switch button to be rendered in the Next.js app

Code changes
======
- Refactors `<ScriptLink />` component to simplify variant switching
- Updates tests to reflect simpler implementation
- Removes `path-to-regexp` library

Testing
======
1. Visit http://localhost:7080/serbian/cyr/srbija-68707945?renderer_env=live
2. Hover over the script switch button and confirm the url changes `cyr` to `lat`
3. Click the link and confirm it takes you to the correct transliterated article (you will need to re-add `?renderer_env=live`)
4. Visit http://localhost:7080/serbian/cyr/srbija-68707945.amp?renderer_env=live
5. Repeat the steps above for AMP (the link will have `.amp` removed, this is expected behaviour)
6. Visit http://localhost:7080/serbian/cyr/srbija-68707945.lite?renderer_env=live
7. Repeat the steps above for Lite
8. Visit http://localhost:7080/serbian/cyr?renderer_env=live
9. Hover over the script switch button and confirm the url changes `cyr` to `lat`
10. Click the link and confirm it takes you to the correct transliterated homepage (you will need to re-add `?renderer_env=live`)
11. Visit http://localhost:7080/serbian/articles/cy9j40v4n75o/cyr?renderer_env=live
12. Hover over the script switch button and confirm the url changes `cyr` to `lat`
13. Click the link and confirm it takes you to the correct transliterated article (you will need to re-add `?renderer_env=live`)

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

amoore108 committed:
WSTEAMA-1558 - Script switch refactor

amoore108 committed:
ScriptLink test suite update

amoore108 committed:
Remove `path-to-regexp`

amoore108 committed:
Update index.tsx

amoore108 committed:
Bundle size update

amoore108 commented on src/server/index.jsx:
```diff
@@ -205,7 +205,7 @@ server.get(
         route: { getInitialData, pageType },
         variant,
-      } = getRouteProps(urlPath);
+      } = getRouteProps(url);
```
Not sure if this is really needed. This is just to maintain query params when using the script switch button, as this value eventually sets the `RequestContext.pathname` value.

amoore108 replied:
This actually mirrors Next.js though. We use `context.resolvedUrl` there which includes the entire URL structure (query params etc) and pass that to `RequestContext.pathname`

amoore108 replied:
Changed this, so no longer applicable

amoore108 committed:
Remove test for variant topic ID as `scriptSwitchId` is now removed

karinathomasbbc commented on src/app/routes/utils/fetchPageData/utils/getRouteProps/index.js:
```diff
@@ -22,7 +22,8 @@ const getNonCanonicalPlatformId = params => {
 
 const getRouteProps = url => {
-  const matchedRoutes = matchRoutes(routes, url);
+  const urlWithoutQuery = url.split('?')?.[0];
```
Should we consider using [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) instead of this? Something along the lines of:

```suggestion
  const { pathname: urlWithoutQuery } = new URL(`https://www.bbc.com${url}`);
```

amoore108 committed:
Fix server url test

amoore108 committed:
Handle `.amp` case

- We want to redirect people from AMP to canonical, so remove `.amp` extension

amoore108 committed:
Merge branch 'latest' into WSTEAMA-1558-script-switch-refactor

amoore108 committed:
Update index.test.tsx

amoore108 committed:
Update index.tsx

karinathomasbbc commented on src/app/components/Header/ScriptLink/index.tsx:
```diff
@@ -1,84 +1,30 @@
-      data-variant={variant}
+      href={
+        pathname
+          .replace(currentVariant as string, alternateVariant)
```
Playing devil's advocate here... what if we had an optimo or tipo ID which starts with `cyr` - there's a risk that it would be replaced by `lat`

How can we make this more robust and less error prone?

amoore108 replied:
Good point. I'll have a think.

karinathomasbbc replied:
You may have to split the URL into parts on `/`, find the one which matches the variant exactly and replace it instead

amoore108 replied:
Added this https://github.com/bbc/simorgh/pull/12328/commits/7534a7f94a17956644d20c2356cf771fe83c510a.

Was hoping `URL` could break the url down into parts, but needs to be done manually.

Thanks for highlighting this though.

karinathomasbbc replied:
Thanks for fixing - nice solution!

amoore108 committed:
Don't maintain query params

- Could lead to a scenario whereby a user on a paginated page switches the script to a page that doesn't exist. So its best to get rid of the query params on script change.

amoore108 committed:
cache

amoore108 committed:
Ensure only variant path part is changed

amoore108 committed:
Update index.test.jsx

amoore108 committed:
Update index.tsx

amoore108 committed:
Update index.tsx

amoore108 committed:
Hide script switch button on Uzbek topics pages

amoore108 committed:
Temporarily hide the script switch button Topic pages

amoore108 committed:
Merge branch 'latest' into WSTEAMA-1558-script-switch-refactor

amoore108 committed:
Merge branch 'latest' into WSTEAMA-1558-script-switch-refactor

amoore108 committed:
Merge branch 'latest' into WSTEAMA-1558-script-switch-refactor

amoore108 committed:
Merge branch 'latest' into WSTEAMA-1558-script-switch-refactor

amoore108 committed:
Show script switch button again

amoore108 committed:
Merge branch 'latest' into WSTEAMA-1558-script-switch-refactor
